add gev distribution and test

blacksde opened the pull request:
Resolve issue #874

blacksde commented:
@srvasude @brianwa84 Could you help review this pr? Thx~

brianwa84 commented:
Pavel, could you take a look?

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+    that supports broadcasting (e.g. `loc + scale` + `concentration` is valid).
+
+    Args:
+      loc: Floating point tensor, the means of the distribution(s).
```
It's not the actual mean. Maybe just call it "the location parameter of the distribution(s)"

blacksde replied:
Thx for pointing this out, I will also fix all other places of misusing mean and loc.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+          parameters=parameters,
+          name=name)
+
+  @staticmethod
```
`_param_shapes` and `_param_event_ndims` methods, we have a new mechanism for describing this. In this case, you can do:

```python
  @classmethod
  def _parameter_properties(cls, dtype, num_classes=None):
    # pylint: disable=g-long-lambda
    return dict(
        loc=parameter_properties.ParameterProperties(),
        scale=parameter_properties.ParameterProperties(
            default_constraining_bijector_fn=(
                lambda: softplus_bijector.Softplus(low=dtype_util.eps(dtype)))))
        concentration=parameter_properties.ParameterProperties(),
    # pylint: enable=g-long-lambda
```

This new style lets us automatically constrain the parameters to the distribution.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+
+  def _entropy(self):
+    # Use broadcasting rules to calculate the full broadcast sigma.
+    scale = self.scale * tf.ones_like(self.loc)
```
Do `tf.broadcast_to(self.scale, ps.broadcast_shape(ps.shape(self.scale), ps.shape(self.loc))` instead, it's a little more self-descriptive (you can drop the comment) and doesn't waste flops.

`ps` should be defined as `from tensorflow_probability.python.internal import prefer_static as ps` up top.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+
+    return self.loc + self.scale * mode_z
+
+  def _default_event_space_bijector(self):
```
Drop this entirely, the superclass has a good default implementation for this distribution (it'll use the composition of sigmoid + GevCDF). Identity is particularly bad, since this distribution often isn't supported on the entire real line.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+  def _entropy(self):
+    # Use broadcasting rules to calculate the full broadcast sigma.
+    scale = self.scale * tf.ones_like(self.loc)
+    return 1. + tf.math.log(scale) + np.euler_gamma*(1. + self.concentration)
```
- Please add spaces around binary operators
- We try our best to avoid numpy type upcasting, so please write this as `tf.constant(np.euler_gamma, self.dtype)`

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+      conc = tf.convert_to_tensor(self.concentration)
+      equal_zero = tf.equal(conc, 0.)
+      log_t = tf.where(equal_zero, -z,
+                       -tf.math.log1p(z * conc) / conc)
```
This needs to be written somewhat differently, because as is it'll produce NaN gradients when conc is 0, see https://github.com/tensorflow/probability/blob/master/discussion/where-nan.pdf for details. In this case, what you can do it something like:

```python
safe_conc = tf.where(equal_zero, 1., conc)
log_t = tf.where(equal_zero, -z,
        -tf.math.log1p(z * safe_conc) / safe_conc)
```

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+    equal_zero = tf.equal(conc, 0.)
+    less_than_one = tf.less(conc, 1.)
+
+    mean_z = tf.where(equal_zero,
```
Same comment about NaN gradients as above (with the same solution, I imagine, although maybe conc=1 isn't a good "safe value").

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+    less_than_one = tf.less(conc, 1.)
+
+    mean_z = tf.where(equal_zero,
+                      np.euler_gamma*tf.ones_like(conc),
```
This can be `tf.fill(tf.shape(conc), tf.constant(np.euler_gamma, self.dtype))`

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+                      np.euler_gamma*tf.ones_like(conc),
+                      tf.where(less_than_one,
+                               tf.math.expm1(tf.math.lgamma(1. - conc)) / conc,
+                               np.inf*tf.ones_like(conc))
```
Same suggestion about using `tf.fill` instead.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+    g1_square = tf.exp(tf.math.lgamma(1. - conc)) ** 2
+    g2 = tf.exp(tf.math.lgamma(1. - 2.*conc))
+
+    std_z = tf.where(equal_zero,
```
Same comment about `NaN` gradients.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+    less_than_half = tf.less(conc, 0.5)
+
+    g1_square = tf.exp(tf.math.lgamma(1. - conc)) ** 2
+    g2 = tf.exp(tf.math.lgamma(1. - 2.*conc))
```
Spaces around binary ops, please.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev.py:
```diff
@@ -0,0 +1,265 @@
+    equal_zero = tf.equal(conc, 0.)
+
+    # Use broadcasting rules to calculate the full broadcast sigma.
+    mode_z = tf.where(equal_zero,
```
Same comment about `NaN` gradients.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev_test.py:
```diff
@@ -0,0 +1,330 @@
+tfd = tfp.distributions
+
+
+class _GEVTest(object):
```
Since I had all those concerns about NaN gradients, would you mind adding a test for those? See https://github.com/tensorflow/probability/blob/612a9cc27cc96f6dad433d604c8255eb9daf5e2e/tensorflow_probability/python/distributions/generalized_normal_test.py#L168-L186 for an example. I especially care about the branch points of concentration=1, 0.5, 0 but other spots may be good to check too.

SiegeLordEx commented on tensorflow_probability/python/distributions/gev_test.py:
```diff
@@ -0,0 +1,330 @@
+
+
+class _GEVTest(object):
+
```
Would you also mind adding a test for broadcasting? Something like this: https://github.com/tensorflow/probability/blob/612a9cc27cc96f6dad433d604c8255eb9daf5e2e/tensorflow_probability/python/distributions/multivariate_student_t_test.py#L36-L108

SiegeLordEx reviewed (changes requested):
Looks good, thanks! I glanced over the math, it seemed okay. I mainly have concerns about the gradients in a few spots.

blacksde commented:
> Looks good, thanks! I glanced over the math, it seemed okay. I mainly have concerns about the gradients in a few spots.

Hi @SiegeLordEx , Thanks a lot for all your valuable comments. I have already fixed all of them(although I didn't reply one by one inline :) ). There are two places to point out:

1. for the broadcasting test case you suggested, I referred to the one in half_student_t_test as its test case class is the same as what I used previously. I don't want to change the basic structure too much so I was hoping this one would be OK.

2. for the gradient NaN issue, I also updated some logic in gev_cdf as the cdf of gev relies on this bijector.

Let me know if there is anything else needed to be updated.

SiegeLordEx commented:
Looks good, thanks. I'll send this along to our internal review and that'll get this merged.
